src/System.Text.Json/tests/Serialization/PropertyNameTests.cs

```diff
@@ -0,0 +1,288 @@
+
+            {
+                // Baseline comparison - no options set.
+                ObjectPropertyNamesDifferentByCaseOnly_TestClass obj = JsonSerializer.Parse<ObjectPropertyNamesDifferentByCaseOnly_TestClass>("{}");
+                JsonSerializer.ToString(obj);
+            }
+
+            {
+                var options = new JsonSerializerOptions();
+                options.PropertyNamingPolicy = JsonNamingPolicy.CamelCase;
+
+                Assert.Throws<InvalidOperationException>(() => JsonSerializer.Parse<ObjectPropertyNamesDifferentByCaseOnly_TestClass>("{}", options));
+                Assert.Throws<InvalidOperationException>(() => JsonSerializer.ToString(new ObjectPropertyNamesDifferentByCaseOnly_TestClass(), options));
+            }
+        }
+
+        [Fact]
+        public static void JsonNameConflictOnCaseInsensitiveFail()
+        {
+            string json = @"{""myInt"":1,""MyInt"":2}";
+
+            {
+                var options = new JsonSerializerOptions();
+                options.PropertyNameCaseInsensitive = true;
+
+                Assert.Throws<InvalidOperationException>(() => JsonSerializer.Parse<IntPropertyNamesDifferentByCaseOnly_TestClass>(json, options));
+                Assert.Throws<InvalidOperationException>(() => JsonSerializer.ToString(new IntPropertyNamesDifferentByCaseOnly_TestClass(), options));
+            }
+        }
+
+        [Fact]
+        public static void JsonOutputNotAffectedByCasingPolicy()
+        {
+            {
+                // Baseline.
+                string json = JsonSerializer.ToString(new SimpleTestClass());
+                Assert.Contains(@"""MyInt16"":0", json);
+            }
+
+            // The JSON output should be unaffected by PropertyNameCaseInsensitive.
+            {
+                var options = new JsonSerializerOptions();
+                options.PropertyNameCaseInsensitive = true;
+
+                string json = JsonSerializer.ToString(new SimpleTestClass(), options);
+                Assert.Contains(@"""MyInt16"":0", json);
+            }
+        }
+
+        [Fact]
+        public static void EmptyPropertyName()
+        {
+            string json = @"{"""":1}";
+
+            {
+                var obj = new EmptyPropertyName_TestClass();
+                obj.MyInt1 = 1;
+
+                string jsonOut = JsonSerializer.ToString(obj);
+                Assert.Equal(json, jsonOut);
+            }
+
+            {
+                EmptyPropertyName_TestClass obj = JsonSerializer.Parse<EmptyPropertyName_TestClass>(json);
+                Assert.Equal(1, obj.MyInt1);
+            }
+        }
+    }
+
+    public class OverridePropertyNameDesignTime_TestClass
+    {
+        [JsonPropertyName("Blah")]
+        public int myInt { get; set; }
+
+        [JsonPropertyName("BlahObject")]
+        public object myObject { get; set; }
+    }
+
+    public class DuplicatePropertyNameDesignTime_TestClass
+    {
+        [JsonPropertyName("Blah")]
```

Comment:
Add a test where camel-casing causes the conflict.

Reply:
See JsonNameConflictOnCamelCasingFail